sdk/rust/oak/src/node.rs

```diff
@@ -0,0 +1,84 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::OakStatus;
+use oak_abi::{label::Label, proto::oak::application::NodeConfiguration};
+pub use oak_io::{Decodable, Encodable, Receiver, Sender};
+
+/// Convenience struct for building Nodes.
+pub struct NodeBuilder {
+    config: NodeConfiguration,
+    label: Option<Label>,
+}
+
+// TODO(#1584): Send init messages as a part of building process.
+impl NodeBuilder {
+    pub fn new(config: &NodeConfiguration) -> Self {
+        Self {
+            config: config.clone(),
+            label: None,
+        }
+    }
+
+    /// Set `label` for both newly created Node and a corresponding initial channel to this Node.
+    pub fn label<'a>(&'a mut self, label: &Label) -> &'a mut Self {
+        self.label = Some(label.clone());
+        self
+    }
+
+    /// Build Node using provided [`NodeBuilder::config`] and [`NodeBuilder::label`].
+    /// If [`NodeBuilder::label`] is `None` - uses [`Label::public_untrusted`].
+    /// Returns [`Sender`] for the initial channel to the newly created Node.
+    pub fn build<T: Encodable + Decodable>(&self) -> Result<Sender<T>, OakStatus> {
+        let (sender, receiver) = match &self.label {
+            Some(label) => {
+                let (sender, receiver) = crate::io::channel_create_with_label::<T>(label)?;
+                crate::node_create_with_label(&self.config, label, receiver.handle)?;
+                (sender, receiver)
+            },
+            None => {
+                let (sender, receiver) = crate::io::channel_create::<T>()?;
+                crate::node_create(&self.config, receiver.handle)?;
+                (sender, receiver)
+            }
+        };
+        crate::channel_close(receiver.handle.handle)?;
+        Ok(sender)
+    }
+}
+
+/// Method provider for creating Wasm Nodes.
+pub struct WasmNode<T: Encodable + Decodable> {
+    phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: Encodable + Decodable> WasmNode<T> {
+    /// Creates a Wasm Node with a public label.
+    /// Returns [`Sender`] for the initial channel to the newly created Node.
+    pub fn create(module_name: &str, entrypoint_name: &str) -> Result<Sender<T>, OakStatus> {
+        Self::create_with_label(module_name, entrypoint_name, &Label::public_untrusted())
+    }
+
+    /// Creates a Wasm Node with specific `label`.
+    /// Returns [`Sender`] for the initial channel to the newly created Node.
+    pub fn create_with_label(module_name: &str, entrypoint_name: &str, label: &Label) -> Result<Sender<T>, OakStatus> {
+        let config = &crate::node_config::wasm(module_name, entrypoint_name);
+        let sender = NodeBuilder::new(config)
+            .label(label)
+            .build::<T>()?;
+        Ok(sender)
+    }
+}
```

Comment:
This seems specific to wasm, I don't understand why it is so coupled to it though, it seems that we would need to copy paste it for every node type, just making it even more coupled, just for the sake of passing module_name and entrypoint_name alongside label.

I don't see the problem with just calling something like `create_node(node::config::wasm(module_name, entrypoint_name), label)` vs  `create_wasm_node(module_name, entrypoint_name, label)`; the latter seems to just take something that is already well structured, and coupling it with another arbitrary function, which if anything, it just breaks composability and encapsulation, just to save a few keystrokes. Or is there any more concrete benefit in the approach you are suggesting @ipetr0v ?

Reply:
We already have `node_create` in SDK - this is why I thought we could add a Node builder
Also, builder could be then extended with initializing Nodes (https://github.com/project-oak/oak/issues/1584) and also working with different types of Nodes:
- Nodes that only listen to incoming requests: `GrpcClientNode`
- Nodes that only send data back: `GrpcServerNode`
- Nodes that both listen and send data back: ones that use the same channel to listen and to send responses back

Reply:
Deleted `WasmNode` and using the Builder itself everywhere